### PR TITLE
Umansky anisotropic diffusion miniapps [umansky-dev]

### PR DESCRIPTION
**Pull request by mlstowell**
Adds a pair of simple plasma miniapps which solve a test problem defined in a paper by Maxim Umansky et al. The test problem is challenging for its use of a discontinuous boundary condition but it primarily studies highly anisotropic diffusion.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#5265](https://github.com/mfem/mfem/pull/5265) | @mlstowell | @tzanio | @najlkin + @psocratis | 3/25/26 | ⌛due 4/8/26 | ⌛due 4/15/26 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: miniapps/plasma/diffusion/umansky_cg.cpp [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+// The test problem consists of a heat equation with an anisotropic
+// diffusion coefficient. The anisotropy is defined by an anisotropy
+// ratio and a direction along which the diffusion differs from
+// unity. In this case the domain is a 2D rectangle and the direction
+// defining the anisotropy is aligned with the main diagonal of the
+// domain. There is no heat source but there is a discontinuous
+// boundary condition with the temperature on the left and upper
+// boundaries equal to one and zero on the remaining boundaries.
+//
+// There are two main challenges presented by this test
```
Comment (najlkin) [inline miniapps/plasma/diffusion/umansky_cg.cpp:31]: It would be better to explain the purpose of the test first, i.e., measurement of artificial diffusion.

File: miniapps/plasma/diffusion/umansky_cg.cpp [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+                  "Diffusion Coefficient Ratio.");
+   args.AddOption(&nc, "-nc", "--non-conforming-amr", "-c",
+                  "--conforming-amr",
+                  "Enable or disable conforming adaptive mesh refinement.");
+   args.AddOption(&device_config, "-d", "--device",
+                  "Device configuration string, see Device::Configure().");
+#ifdef MFEM_USE_CEED
+   args.AddOption(&algebraic_ceed, "-a", "--algebraic",
+                  "-no-a", "--no-algebraic",
+                  "Use algebraic Ceed solver");
```
Comment (najlkin) [inline miniapps/plasma/diffusion/umansky_cg.cpp]: Not used.

File: miniapps/plasma/diffusion/umansky_cg.cpp [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+                  "Number of elements in y direction.");
+   args.AddOption(&w, "-wth", "--width",
+                  "Width of domain.");
+   args.AddOption(&h, "-hgt", "--height",
+                  "Height of domain.");
+   args.AddOption(&Ak, "-Ak", "--diff-ratio",
+                  "Diffusion Coefficient Ratio.");
+   args.AddOption(&nc, "-nc", "--non-conforming-amr", "-c",
+                  "--conforming-amr",
+                  "Enable or disable conforming adaptive mesh refinement.");
```
Comment (najlkin) [inline miniapps/plasma/diffusion/umansky_cg.cpp]: What was the idea here? Using triangles instead of quads? These options can be tied together as it does not make much sense on its own and just gives ugly errors trying to use conforming AMR with quads.

File: miniapps/plasma/diffusion/umansky_cg.cpp [hunk 4]
```diff
@@ -0,0 +1,435 @@
...
+      pmesh.UniformRefinement();
+   }
+   // Make sure tet-only meshes are marked for local refinement.
+   pmesh.Finalize(true);
+
+   // Define a parallel finite element space on the parallel mesh. Here we
+   // use continuous Lagrange finite elements of the specified order. If
+   // order < 1, we instead use an isoparametric/isogeometric space.
+   FiniteElementCollection *fec;
+   bool delete_fec;
```
Comment (najlkin) [inline miniapps/plasma/diffusion/umansky_cg.cpp]: A unique_ptr could be used instead

File: miniapps/plasma/diffusion/umansky_cg.cpp [hunk 5]
```diff
@@ -0,0 +1,435 @@
...
+      args.PrintOptions(cout);
+   }
+
+   // Enable hardware devices such as GPUs, and programming models such as
+   // CUDA, OCCA, RAJA and OpenMP based on command line options.
+   Device device(device_config);
+   if (myid == 0) { device.Print(); }
+
+
+
```
Comment (najlkin) [inline miniapps/plasma/diffusion/umansky_cg.cpp]: Delete extra empty lines.

File: miniapps/plasma/diffusion/umansky_cg.cpp [hunk 6]
```diff
@@ -0,0 +1,435 @@
...
+   // Determine the list of true (i.e. parallel conforming) essential
+   // boundary dofs. In this example, the boundary conditions are defined
+   // by marking all the external boundary attributes from the mesh as
+   // essential (Dirichlet) and converting them to a list of true dofs.
+   Array<int> ess_tdof_list;
+   Array<int> ess_bdr(pmesh.bdr_attributes.Max());
+   if (pmesh.bdr_attributes.Size())
+   {
+      ess_bdr = 1;
+   }
```
Comment (najlkin) [inline miniapps/plasma/diffusion/umansky_cg.cpp]: The problem is not setting value of the array (which can be empty), but calling `Max()` on an empty array.

File: miniapps/plasma/diffusion/umansky_cg.cpp
Comment (najlkin) [inline miniapps/plasma/diffusion/umansky_cg.cpp:1]: In general, why not to use the code for measurement of the transition width? It is nice to watch the process of AMR, but it does not give any quantifiable result to compare between methods or with a paper...

File: miniapps/plasma/diffusion/umansky.hpp [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+   void Eval(DenseMatrix &M, ElementTransformation &T,
+             const IntegrationPoint &ip)
+   {
+      M.SetSize(2);
+
+      M(0,0) = 1.0 + (Ak_ - 1.0) * w_ * w_ / d2_;
+      M(0,1) = (Ak_ - 1.0) * w_ * h_ / d2_;
+      M(1,0) = M(0,1);
+      M(1,1) = 1.0 + (Ak_ - 1.0) * h_ * h_ / d2_;
```
Comment (najlkin) [inline miniapps/plasma/diffusion/umansky.hpp:105]: Explanation in terms of goniometric functions would be nice here to match it with the paper.

File: miniapps/plasma/diffusion/umansky.hpp [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+   real_t Eval(ElementTransformation &T,
+               const IntegrationPoint &ip)
+   {
+      real_t x[2];
+      Vector transip(x, 2);
+
+      T.Transform(ip, transip);
+
+      if (fabs(w_ * x[1] - h_ * x[0]) < 1e-6 * sqrt(w_ * h_))
```
Comment (najlkin) [inline miniapps/plasma/diffusion/umansky.hpp]: The hardcoded value here is not ideal, does not it cause problems for fine refinement? 🤔

File: miniapps/plasma/diffusion/umansky.hpp [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_UMANSKY_HPP
+#define MFEM_UMANSKY_HPP
+
+#include "mfem.hpp"
+
+using namespace mfem;
+
+class UmanskyBoundaryCoefficient : public Coefficient
```
Comment (najlkin) [inline miniapps/plasma/diffusion/umansky.hpp]: Some short description would be good.

File: miniapps/plasma/diffusion/umansky_cg.cpp [hunk 10]
```diff
@@ -0,0 +1,435 @@
...
+      // Send the solution by socket to a GLVis server if enabled.
+      if (visualization)
+      {
+         sol_sock << "parallel " << num_procs << " " << myid << "\n";
+         sol_sock.precision(8);
+         sol_sock << "solution\n" << pmesh << x
+                  << " window_title 'Number of DoFs: " << prob_size << "'";
+         if (it == 1)
+         {
+            sol_sock << " keys 'mmjR'\n";
```
Comment (najlkin) [inline miniapps/plasma/diffusion/umansky_cg.cpp:390]:
```suggestion
            sol_sock << " keys 'mmjRl'\n";
```
The default light source does not show the transition well.

File: miniapps/plasma/diffusion/umansky_cg.cpp [hunk 11]
```diff
@@ -0,0 +1,435 @@
...
+
+      if (c != 'c')
+      {
+         break;
+      }
+
+      // Estimate element errors using the Zienkiewicz-Zhu error estimator.
+      Vector errors(pmesh.GetNE());
+      {
+         DiffusionIntegrator flux_integrator(anisoDiffCoef);
```
Comment (najlkin) [inline miniapps/plasma/diffusion/umansky_cg.cpp:437]: The integrator could be taken from the bilinear form for consistency.

File: miniapps/plasma/diffusion/umansky_cg.cpp [hunk 12]
```diff
@@ -0,0 +1,435 @@
...
+// domain. Second, as the anisotropy ratio is increased the numerical
+// scheme may introduce anomalous diffusion in the direction
+// perpendicular to the highly diffusive direction.
+//
+// At sufficiently high anisotropy ratios the solver may fail to find
+// an acceptable solution. Therefore, the purpose of this mini
+// application is to provide a starting point for evaluating solution
+// methods based on a continuous nodal discretization of the
+// temperature field.
+//
```
Comment (najlkin) [inline miniapps/plasma/diffusion/umansky_cg.cpp:43]: AMR procedure could be mentioned and reference to examples.

File: miniapps/plasma/diffusion/umansky_cg.cpp [hunk 13]
```diff
@@ -0,0 +1,435 @@
...
+
+         L2ZZErrorEstimator(flux_integrator, x,
+                            smooth_flux_fes, flux_fes, errors, norm_p);
+
+      }
+
+      real_t local_max_err = errors.Max();
+      real_t global_max_err;
+      MPI_Allreduce(&local_max_err, &global_max_err, 1,
+                    MPITypeMap<real_t>::mpi_type, MPI_MAX, pmesh.GetComm());
```
Comment (najlkin) [inline miniapps/plasma/diffusion/umansky_cg.cpp]:
```suggestion
                    MFEM_MPI_REAL_T, MPI_MAX, pmesh.GetComm());
```
Matter of personal preference, I guess...

File: miniapps/plasma/diffusion/umansky_dg.cpp [hunk 14]
```diff
@@ -0,0 +1,452 @@
...
+                                                         sigma, kappa));
+   a.AddBdrFaceIntegrator(new DGDiffusionIntegrator(anisoDiffCoef,
+                                                    sigma, kappa));
+   if (eta > 0)
+   {
+      MFEM_VERIFY(!pa, "BR2 not yet compatible with partial assembly.");
+      a.AddInteriorFaceIntegrator(new DGDiffusionBR2Integrator(fespace, eta));
+      a.AddBdrFaceIntegrator(new DGDiffusionBR2Integrator(fespace, eta));
+   }
+   if (pa) { a.SetAssemblyLevel(AssemblyLevel::PARTIAL); }
```
Comment (najlkin) [inline miniapps/plasma/diffusion/umansky_dg.cpp]: Partial assembly is not very meaningful with AMR, it can only run with quads without any refinement.

File: miniapps/plasma/diffusion/umansky_dg.cpp [hunk 15]
```diff
@@ -0,0 +1,452 @@
...
+   a.AddInteriorFaceIntegrator(new DGDiffusionIntegrator(anisoDiffCoef,
+                                                         sigma, kappa));
+   a.AddBdrFaceIntegrator(new DGDiffusionIntegrator(anisoDiffCoef,
+                                                    sigma, kappa));
+   if (eta > 0)
+   {
+      MFEM_VERIFY(!pa, "BR2 not yet compatible with partial assembly.");
+      a.AddInteriorFaceIntegrator(new DGDiffusionBR2Integrator(fespace, eta));
+      a.AddBdrFaceIntegrator(new DGDiffusionBR2Integrator(fespace, eta));
+   }
```
Comment (najlkin) [inline miniapps/plasma/diffusion/umansky_dg.cpp]: BR2 also does not work with AMR.

File: miniapps/plasma/diffusion/umansky_dg.cpp [hunk 16]
```diff
@@ -0,0 +1,452 @@
...
+
+      if (c != 'c')
+      {
+         break;
+      }
+
+      // Estimate element errors using the Zienkiewicz-Zhu error estimator.
+      Vector errors(pmesh.GetNE());
+      {
+         DiffusionIntegrator flux_integrator(anisoDiffCoef);
```
Comment (najlkin) [inline miniapps/plasma/diffusion/umansky_dg.cpp:444]: This construction should be explained here or in the header of the file. It is not ideal to use CG error estimator for DG, but the mesh is not aligned with the discontinuity, so it is not the worst here...

**Review by najlkin — COMMENTED** [review]
Master should be remerged, mainly because of #5319 to not cause the same problem with double building plasma miniapps.

**Comment by najlkin** [comment]
Regarding the general discussion about merging the files, it is connected to resolving the comments above. If the options of the DGs which are not compatible with AMR are removed, there will be just a very few differences between the CG and DG versions, so it would make more sense to merge them 🤔 .

File: miniapps/plasma/diffusion/umansky.hpp [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+      M(0,1) = (Ak_ - 1.0) * w_ * h_ / d2_;
+      M(1,0) = M(0,1);
+      M(1,1) = 1.0 + (Ak_ - 1.0) * h_ * h_ / d2_;
+   }
+
+private:
+   real_t w_, h_, Ak_, d2_;
+};
+
+#endif // MFEM_UMANSKY_SOLVER
```
Comment (psocratis) [inline miniapps/plasma/diffusion/umansky.hpp]: should be `MFEM_UMANSKY_HPP`

**Comment by mlstowell** [comment]
> Regarding the general discussion about merging the files, it is connected to resolving the comments above. If the options of the DGs which are not compatible with AMR are removed, there will be just a very few differences between the CG and DG versions, so it would make more sense to merge them 🤔 .

I ran into trouble during the development of the Stix miniapps because merging different capabilities into a single code made some people nervous that Stix would compete with their more general application. I am therefore reluctant to create a more general miniapp even if it makes perfect sense. It has now been more than a year since this issue arose with Stix and we still don't have a Stix miniapp. Frankly, I would rather abandon this entire project than revisit this choice regarding the balance between generality and actually getting something into MFEM. In the future someone may decide to combine these simple codes into a more general code and I would welcome them to do so but I have absolutely no interest in being that person.

File: miniapps/plasma/diffusion/umansky_cg.cpp (hunk 1 above)
Comment (mlstowell) [reply miniapps/plasma/diffusion/umansky_cg.cpp:31]: That is a matter of opinion. The code does not actually produce a measurement of the artificial diffusion so it would not be appropriate to claim that this is the purpose of the code. I agree that such a measure would be useful but this code only attempts to provide qualitative comparison to other methods.

File: miniapps/plasma/diffusion/umansky_cg.cpp (hunk 2 above)
Comment (mlstowell) [reply miniapps/plasma/diffusion/umansky_cg.cpp]: Good catch

File: miniapps/plasma/diffusion/umansky_cg.cpp (hunk 3 above)
Comment (mlstowell) [reply miniapps/plasma/diffusion/umansky_cg.cpp]: This is lifted directly from ex6p.cpp. I am looking forward to the day when MFEM's mixed mesh AMR capabilities are more general. I can take it out.

File: miniapps/plasma/diffusion/umansky_dg.cpp (hunk 14 above)
Comment (mlstowell) [reply miniapps/plasma/diffusion/umansky_dg.cpp]: Should I remove the option, add a comment, or leave it for people who want to experiment?

File: miniapps/plasma/diffusion/umansky_dg.cpp (hunk 15 above)
Comment (mlstowell) [reply miniapps/plasma/diffusion/umansky_dg.cpp]: This was lifted from ex14p.cpp. I can remove it or leave it for people who might want to experiment. I have no preference.

File: miniapps/plasma/diffusion/umansky_cg.cpp (hunk 1 above)
Comment (najlkin) [reply miniapps/plasma/diffusion/umansky_cg.cpp:31]: Why not to quantitatively measure it when you have the code for that already? There would be at least some good use of it. It is hard to compare the transition width visually, even when we know it is not very representative for high-order finite elements, but it does not hurt to provide to the user, I think.

Comment (mlstowell) [reply miniapps/plasma/diffusion/umansky_cg.cpp:31]: Perhaps I have forgotten. Where would I find this code?

File: miniapps/plasma/diffusion/umansky_cg.cpp (hunk 4 above)
Comment (mlstowell) [reply miniapps/plasma/diffusion/umansky_cg.cpp]: This was lifted from ex1p.cpp. I'm not familiar enough with unique_ptr to trust that I would use it correctly.

File: miniapps/plasma/diffusion/umansky_dg.cpp (hunk 16 above)
Comment (mlstowell) [reply miniapps/plasma/diffusion/umansky_dg.cpp:444]: I'm not an expert on such things. I'd be happy to use a more appropriate estimator if there is a straightforward option.

File: miniapps/plasma/diffusion/umansky_cg.cpp (hunk 3 above)
Comment (najlkin) [reply miniapps/plasma/diffusion/umansky_cg.cpp]: I see it saying there "--nonconforming-simplices", which suggests it is meant for simplices. For quads, you do not have that option really as there is no meaningful way how to do conforming AMR. So I would just put that `EnsureNC()` in an `if` for the type of the elements.

File: miniapps/plasma/diffusion/umansky_cg.cpp (hunk 4 above)
Comment (najlkin) [reply miniapps/plasma/diffusion/umansky_cg.cpp]:
```suggestion
   unique_ptr<FiniteElementCollection> own_fec;
```
It (`unique_ptr`) is actually used later for BoomerAMG, just do `own_fec = make_unique<H1_FECollection>(order, dim); fec = own_fec.get();`.

File: miniapps/plasma/diffusion/umansky_dg.cpp (hunk 16 above)
Comment (najlkin) [reply miniapps/plasma/diffusion/umansky_dg.cpp:444]: I do not think there is a clear alternative for DGs. That is one of the benefits of HDG, which gives such estimate naturally 🤔 . Anyway, I think it is fine, it should be just mentioned😉 .

File: miniapps/plasma/diffusion/umansky_dg.cpp (hunk 15 above)
Comment (najlkin) [reply miniapps/plasma/diffusion/umansky_dg.cpp]: Depends on the concept of the miniapp, I think. Right now, it is centered at AMR, which is also default. So if it is so, I would remove these options 🤷 .

File: miniapps/plasma/diffusion/umansky_cg.cpp (hunk 1 above)
Comment (najlkin) [reply miniapps/plasma/diffusion/umansky_cg.cpp:31]: `UmanskyTestWidth()` in plasma-dev:miniapps/plasma/transport2d.cpp

File: miniapps/plasma/diffusion/umansky_cg.cpp [hunk 18]
```diff
@@ -0,0 +1,429 @@
...
+   }
+
+   // Define a parallel mesh by a partitioning of the serial mesh. Refine
+   // this mesh further in parallel to increase the resolution. Once the
+   // parallel mesh is defined, the serial mesh can be deleted.
+   ParMesh pmesh(MPI_COMM_WORLD, mesh);
+   mesh.Clear();
+
+   // Refine this mesh in parallel to increase the resolution.
+   int par_ref_levels = parallel_ref_levels;
```
Comment (psocratis) [inline miniapps/plasma/diffusion/umansky_cg.cpp]: this copy can be avoided

File: miniapps/plasma/diffusion/umansky_cg.cpp [hunk 19]
```diff
@@ -0,0 +1,429 @@
...
+            cout << "Reached maximum number of iterations, exiting..." << endl;
+         }
+         break;
+      }
+
+      // Wait for user input. Ask every 10th iteration.
+      char c = 'c';
+      if (Mpi::Root() && (it % 10 == 0))
+      {
+         cout << "press (q)uit or (c)ontinue --> " << flush;
```
Comment (psocratis) [inline miniapps/plasma/diffusion/umansky_cg.cpp:423]: for me this makes the pause to happen before the print and so the user does not see the instruction. 
I tried to use `std::endl` instead of `flush` and it works

File: miniapps/plasma/diffusion/umansky.hpp (hunk 9 above)
Comment (psocratis) [inline miniapps/plasma/diffusion/umansky.hpp]: Is this coefficient really needed? I see that we only use the values on the boundaries. On boundaries 0,1 the value is zero and on boundaries 2,3 the value is one. So in the _cg case you can just project the `ConstantCoefficient one(1.0)` on just those boundaries. Similarly, in the dg case you can pass the boundary marker indicating only the non-zero boundaries in the `BdrFaceIntegrator`

File: miniapps/plasma/diffusion/umansky_cg.cpp [hunk 20]
```diff
@@ -0,0 +1,429 @@
...
+   // Determine the list of true (i.e. parallel conforming) essential
+   // boundary dofs. In this example, the boundary conditions are defined
+   // by marking all the external boundary attributes from the mesh as
+   // essential (Dirichlet) and converting them to a list of true dofs.
+   Array<int> ess_tdof_list;
+   Array<int> ess_bdr(pmesh.bdr_attributes.Max());
+   if (pmesh.bdr_attributes.Size())
+   {
+      ess_bdr = 1;
+   }
```
Comment (psocratis) [inline miniapps/plasma/diffusion/umansky_cg.cpp]:
```suggestion
   Array<int> ess_bdr;
   if (pmesh.bdr_attributes.Size())
   {
      ess_bdr.SetSize(pmesh.bdr_attributes.Max())
      ess_bdr = 1;
   }
```

File: miniapps/plasma/diffusion/umansky_cg.cpp [hunk 21]
```diff
@@ -0,0 +1,429 @@
...
+   visit_dc.SetFormat(DataCollection::PARALLEL_FORMAT);
+
+   if (visit)
+   {
+      visit_dc.RegisterField("solution", &x);
+   }
+
+   // The main AMR loop. In each iteration we solve the problem on the current
+   // mesh, visualize the solution, estimate the error on all elements, refine
+   // the worst elements and update all objects to work with the new mesh.  We
```
Comment (psocratis) [inline miniapps/plasma/diffusion/umansky_cg.cpp]: maybe better ... "refine the elements with the largest estimated error" ...

File: miniapps/plasma/diffusion/umansky_cg.cpp [hunk 22]
```diff
@@ -0,0 +1,429 @@
...
+
+   // Set up the parallel bilinear form a(.,.) on the finite element space
+   // corresponding to the Laplacian operator -Delta, by adding the
+   // Diffusion domain integrator.
+   AnisotropicDiffusionCoefficient anisoDiffCoef(w, h, Ak);
+   ParBilinearForm a(&fespace);
+   a.AddDomainIntegrator(new DiffusionIntegrator(anisoDiffCoef));
+
+   // Initialize VisIt visualization
+   VisItDataCollection visit_dc("Umansky-CG-AMR-Parallel", &pmesh);
```
Comment (psocratis) [inline miniapps/plasma/diffusion/umansky_cg.cpp]: better to use a pointer to avoid the call to the constructor if VisIt is not used.

File: miniapps/plasma/diffusion/umansky_dg.cpp [hunk 23]
```diff
@@ -0,0 +1,446 @@
...
+   }
+
+   // Define a parallel mesh by a partitioning of the serial mesh. Refine
+   // this mesh further in parallel to increase the resolution. Once the
+   // parallel mesh is defined, the serial mesh can be deleted.
+   ParMesh pmesh(MPI_COMM_WORLD, mesh);
+   mesh.Clear();
+
+   // Refine this mesh in parallel to increase the resolution.
+   int par_ref_levels = parallel_ref_levels;
```
Comment (psocratis) [inline miniapps/plasma/diffusion/umansky_dg.cpp]: this copy is not necessary

File: miniapps/plasma/diffusion/umansky.hpp (hunk 9 above)
Comment (mlstowell) [reply miniapps/plasma/diffusion/umansky.hpp]: I think it is necessary for the corner nodes in the CG case where I want the value to be 0.5. Without this I believe the computed solution has a small asymmetry across the diagonal.

File: miniapps/plasma/diffusion/umansky_cg.cpp (hunk 21 above)
Comment (mlstowell) [reply miniapps/plasma/diffusion/umansky_cg.cpp]: Definitely better.

File: miniapps/plasma/diffusion/umansky_cg.cpp (hunk 22 above)
Comment (mlstowell) [reply miniapps/plasma/diffusion/umansky_cg.cpp]: Should I add a ParaView option while I'm at it? I have a bias toward VisIt but I probably should get in the habit of offering more options.

File: miniapps/plasma/diffusion/umansky_cg.cpp (hunk 19 above)
Comment (mlstowell) [reply miniapps/plasma/diffusion/umansky_cg.cpp:423]: Hmm. We use the same thing in ex32p.cpp and very similar code in examples 11, 12, and 13. I'm reluctant to make this the outlier. Do you see similar behavior in those examples? Also, what platform do you see this on?

File: miniapps/plasma/diffusion/umansky_cg.cpp (hunk 22 above)
Comment (psocratis) [reply miniapps/plasma/diffusion/umansky_cg.cpp]: ParaView option would be nice!

File: miniapps/plasma/diffusion/umansky_cg.cpp (hunk 19 above)
Comment (psocratis) [reply miniapps/plasma/diffusion/umansky_cg.cpp:423]: on an M1 Mac ... and yes I just checked ex32 and 11 and I see the same issue. Btw, equivalently if you add "\n" at the end printing string, it also fixes the issue for me

File: miniapps/plasma/diffusion/umansky_cg.cpp (hunk 4 above)
Comment (mlstowell) [reply miniapps/plasma/diffusion/umansky_cg.cpp]: Thanks for the tip

File: miniapps/plasma/diffusion/umansky.hpp (hunk 8 above)
Comment (mlstowell) [reply miniapps/plasma/diffusion/umansky.hpp]: Good point.

File: miniapps/plasma/diffusion/umansky.hpp (hunk 7 above)
Comment (mlstowell) [reply miniapps/plasma/diffusion/umansky.hpp:105]: Another good point

File: miniapps/plasma/diffusion/umansky_cg.cpp
Comment (mlstowell) [reply miniapps/plasma/diffusion/umansky_cg.cpp:1]: Good suggestion. What do you have in mind? I can certainly produce an ascii file with data such as width, number of DoFs, minimum edge length, etc. but that requires users to process the file somehow. Incidentally, I don't believe we can compute the error as they do in the paper because we don't have the semi-analytic solution implemented. I could also produce a custom visualization in GLVis showing some of this data similar to how example 20 displays its results.

**Comment by mlstowell** [comment]
Umansky_CG_Ak_1e3
An example of the type of plot produced by GnuPlot using profile width data from `umansky_cg.cpp`. The bottom plot is similar to those shown in the paper but the other two are also interesting from an AMR point of view.

**Comment by mlstowell** [comment]
> An example of the type of plot produced by GnuPlot using profile width data from `umansky_cg.cpp`. The bottom plot is similar to those shown in the paper but the other two are also interesting from an AMR point of view.

Note that the title of the plot is incorrect because it contains (DG) rather than (CG). This has been corrected in the code.

**Comment by tzanio** [comment]
@copilot+claude-sonnet-4.5, can you double check all items in .github/copilot-instructions.md were addressed in your review above?

**Comment by tzanio** [comment]
Please update `CHANGELOG` (see [CONTRIBUTING.md](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-request-checklist))

File: miniapps/plasma/diffusion/umansky_dg.cpp [hunk 24]
```diff
@@ -0,0 +1,510 @@
...
+      // Estimate element errors using the Zienkiewicz-Zhu error estimator.
+      Vector errors(pmesh.GetNE());
+      {
+         if (Mpi::Root()) { cout << "Estimating error..." << endl; }
+         DiffusionIntegrator flux_integrator(anisoDiffCoef);
+         L2_FECollection flux_fec(order, dim);
+         ParFiniteElementSpace flux_fes(&pmesh, &flux_fec, 2);
+
+         // Space for the smoothed (conforming) flux
+         int norm_p = 1;
```
[the remark this answers is not in the document]
Comment (najlkin) [reply miniapps/plasma/diffusion/umansky_dg.cpp:449]: Oh, this is a good point connected to what I was mentioning about DG AMR. ZZ is not really a good estimator for this and as it does not consider the jump terms, it will fail with piecewise-constant functions 🫠 . As we do not have an alternative, it may just show an error or turn off AMR.

File: miniapps/plasma/diffusion/CMakeLists.txt [hunk 25]
```diff
@@ -0,0 +1,40 @@
...
+
+# Add the corresponding tests to the "test" target
+  if (MFEM_ENABLE_TESTING)
+    add_test(NAME umansky_cg_np=4
+      COMMAND ${MPIEXEC} ${MPIEXEC_NUMPROC_FLAG} ${MFEM_MPI_NP}
+      ${MPIEXEC_PREFLAGS}
+      $<TARGET_FILE:umansky_cg> -no-vis -maxit 2 -rs 1 -rp 1 -hgt 0.8 -wth 1.2 -Ak 1e2
+      ${MPIEXEC_POSTFLAGS})
+
+    add_test(NAME umansky_dg_np=4
```
[the remark this answers is not in the document]
Comment (najlkin) [reply miniapps/plasma/diffusion/CMakeLists.txt]: This actually causes fails in some autotests as they use just two or three processors. But it is just a name here...
